src/crowdsale/CrowdSale.sol

```diff
@@ -43,15 +48,24 @@ contract CrowdSale {
     constructor() { }
 
     function startSale(Sale memory sale) public returns (uint256 saleId) {
+        if (sale.closingTime < block.timestamp + 1 hours) {
+            revert BadSaleDuration();
+        }
+        if (sale.auctionToken.decimals() != 18 || IERC20Metadata(address(sale.biddingToken)).decimals() != 18) {
+            revert BadDecimals();
+        }
         if (sale.auctionToken.balanceOf(msg.sender) < sale.salesAmount) {
-            revert("you dont have sufficient auction tokens");
+            revert BalanceTooLow();
         }
         //close to 0 cases lead to very confusing results
         // if (sale.fundingGoal <= 0.5 ether || sale.salesAmount < 0.5 ether) {
         //     revert("you must sell or accept something meaningful");
         // }
 
         saleId = uint256(keccak256(abi.encode(sale)));
+        if (address(_sales[saleId].auctionToken) != address(0)) {
+            revert SaleAlreadyActive();
+        }
```

Comment:
This might be a confusing error to encounter if a user wants to create a second auction after the first one has been settled with all the same parameteres. 
If we are not calculating the saleId differently then we should give a more descriptive error here like "SaleAlreadyExists" or something. Still not ideal, b/c it should be possible to create two identical crowdSales one after the other.

Reply:
they are *not* identical but differ in their target timestamps.

Reply:
this changeset is too large for me atm, I'm on the frontend. Lets go with this one.